src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,259 @@
+                                                                                                   clientList(_max_clients),
+                                                                                                   runner(&Manager::run, this),
+                                                                                                   stopped(false), cct(_cct),
+                                                                                                   idle_time(std::chrono::milliseconds(idle_time_ms)){}
+    // dtor wait for thread to stop
+    // then connection are cleaned-up
+    ~Manager() {
+      stopped = true;
+      runner.join();
+      messageQueue.consume_all(delete_message);
+    }
+
+
+    AwsClient connect(const std::string &accessKey,
+                      const std::string &accessSecret,
+                      const std::string &arn,
+                      const std::string &caPath,
+                      bool useSSL) {
+      if (stopped) {
+        return AwsClient();
+      }
+      const client_id_t id(accessKey, accessSecret, arn, useSSL);
+      std::lock_guard lock(clientsLock);
+      const auto it = clientList.find(id);
+      if (it != clientList.end()) {
+        return it->second;
+      }
+
+      // connection not found, creating a new one
+      Aws::Client::ClientConfiguration configuration;
+      if (useSSL) {
+        configuration.scheme = Aws::Http::Scheme::HTTPS;
+        configuration.verifySSL = true;
+      } else {
+        configuration.scheme = Aws::Http::Scheme::HTTP;
+        configuration.verifySSL = false;
+      }
+      Aws::Utils::ARN arn1((Aws::String(arn)));
+      configuration.region = Aws::String(arn1.GetRegion());
+      configuration.maxConnections = 25;
+      configuration.requestTimeoutMs = 3000;
+      configuration.connectTimeoutMs = 3000;
+      configuration.caPath = Aws::String(caPath);
+      Aws::Auth::AWSCredentials credentials;
+      credentials.SetAWSAccessKeyId(Aws::String(accessKey));
+      credentials.SetAWSSecretKey(Aws::String(accessSecret));
+      AwsClient client;
+      if (arn1.GetService() == "lambda") {
+        client = new Aws::Lambda::LambdaClient(credentials, configuration);
+      } else if (arn1.GetService() == "sns") {
+        client = new Aws::SNS::SNSClient(credentials, configuration);
+      }
+      return clientList.emplace(id, client).first->second;
+    }
+
+    int publish(AwsClient client,
+                const std::string &payload,
+                const std::string &resourceARN) {
+      if (stopped) {
+        return STATUS_MANAGER_STOPPED;
+      }
+      if (client == AwsClient()) {
+        return STATUS_CLIENT_CLOSED;
+      }
+      if (messageQueue.push(new message_t(client, resourceARN, payload))) {
+        return STATUS_OK;
+      }
+
+
+      return STATUS_FULL;
+    }
+
+    void stop() {
+      stopped = true;
+    }
+
+  };
+
+  static Manager *manager = nullptr;
+  Aws::SDKOptions options;
```

Comment:
why not make `options` a member of the manager?
init and destroy it inside the manager's ctor/dtor

also, what kind of values does `options` get?

Reply:
then, won't the sdk be init multiple times if we create multiple managers, in case

Reply:
yes. it would be init multiple times. is this an issue?
(trying to minimize the amount of singletons we have)

Reply:
ok, I guess different http client, logging option, etc. might be reqd. for different manager instance. so this seems good